### ISSUE TITLE
Scheduler test improvements and forensics. 

### ISSUE THREAD

**Issue by jayunit100** [kubernetes/kubernetes#44342]
*Problem*

Two related problems of scheduler test coverage and forensics are described in this issue. 

1) Scheduling decisions statistics and forensics are either 
- summarized in events or 
- grepped from logs.  

2) There is no way to 'what-if' test various scheduling scenarios (i.e. X predicate set, Y priority set, on Z # of nodes for A # of pods).  

*Current solution* 

Obviously (1) and (2) are commonly implemented by current users.  But how?   Typically in one-off solutions across the codebase that need continual maintenance, and mind-numbing debates about what information we should spam by default.

- This doesn't allow us to provide intuitive, stateful summaries to admins, users - we have to **manually 'rig' logs based on developer intuition on a case-by-case basis** for what we do/do-not want (#44115).  This is prone to be overlooked and easily  can slip through our fingers - relying on ad-hoc logs is not a reliable way to provide users with essential scheduling data. 
- It forces us to make up **ad-hoc summary statistics** in things like scheduler_perf, which also aren't available for regular scheduler users, and aren't easily maintained.  Right now we have heuristic cutoffs for performance regression.
- It also forces us to do polling to get statistics over time: `scheduled, err := schedulerConfigFactory.GetScheduledPodLister().List(labels.Everything()) (in scheduler_bench_test.go)`.  This may be a performance issue for tests at the very least.  At worst its both a performance issue, technical debt, and maybe even an inaccurate data collection which isn't centrally vetted.
- And the ad-hoc statistics are really hard to use for any real analysis:
```
	if min < threshold3K {
		t.Errorf("Failing: Scheduling rate was too low for an interval, we saw rate of %v, which is the allowed minimum of %v ! ", min, threshold3K)
	} else if min < warning3K {
		fmt.Printf("Warning: pod scheduling throughput for 3k pods was slow for an interval... Saw a interval with very low (%v) scheduling rate!", min)
	} else {
		fmt.Printf("Minimal observed throughput for 3k pod test: %v\n", min)
	}
```
- And again in density, we do similar roll ups - not that these should be replaced, but it shows the importance of summary statistics rollups:
```
I0411 17:33:19.405591   21550 runners.go:174] Created deployment with name: density30-0-f406b9bf-1edc-11e7-84c1-42010af00003, namespace: e2e-tests-density-30-1-9b07j, replica count: 30
Apr 11 17:33:29.402: INFO: Density Pods: 30 out of 30 created, 0 running, 30 pending, 0 waiting, 0 inactive, 0 terminating, 0 unknown, 0 runningButNotReady
Apr 11 17:33:39.403: INFO: Density Pods: 30 out of 30 created, 0 running, 30 pending, 0 waiting, 0 inactive, 0 terminating, 0 unknown, 0 runningButNotReady
Apr 11 17:33:49.402: INFO: Density Pods: 30 out of 30 created, 6 running, 24 pending, 0 waiting, 0 inactive, 0 terminating, 0 unknown, 0 runningButNotReady
(the logs are unreadable at this scale)
```
*Data points*

- We've seen potentially 1000s of log statements creeping up in issues like #44115 . 
- We have to obfuscate summary data using things like V(10) : #37558, where you can get O(np) logs.
- A good example of this is the debate around wether or not we want `glog.Info.V(4)` or not by default in large clusters ( #44115 ).
- So, for example, making logging smarter adds technical debt into the core scheduler, also undesireable, i.e. you might make something that looks like this:
```
podsFailureLogged map[string]bool = ...
if ! podsFailureLogged[thisPod] {
    glog.Info("Failed scheduling %v ",thisPod)
    podsFailureLogged[thisPod] = true
}
```

*Solution* 

Almost all the above bullets can be solved,  if the scheduler had a module inside of it that handled forensics.  

In the said module, we can have: 
- Stateful rollups with intuitive info.
- Unliterreed algorithm code that doesn't need tie various scheduling decisions into the overall log.
- Unspammy forensics which do the roll up for you rather then forcing you to do it yourself.
- A natural and portable mechanism for comparing different scheduler performance runs.

For example, we could have summary data such as

```
1:00:00:20 In the last minute, X pods have been scheduled, with  Y failed scheduling events.
1:00:00:20 Z Scheduled pods with the prefix 'abcd' failed scheduling for the first time.
1:00:00:20 Fit function performance outliers: Predicate X takes Y% of the time on the cluster.  
1:00:00:20 Unique Nodes scheduled in the past minute: 100.
...  
... 
and so on.
``` 

*Alternatives*

- Grep scheduling decisions from logs / events.  This obviously doesnt scale.
- Use metrics to 'reverse engineer' changing scheduler stats over time.  Reasonable but requires time series data wrangling - which isn't something we should need to do in order to measure scheduling performance.

*Implementation options*

- Fork logging so that some logs are sent to the forensics module which cacluates and dumps summaries to another glog or logrus directory path.  This is probably the easiest, most straight forward implementation.
- Maybe update the scheduler API to publish these summaries to a REST endpoint periodically ? That might be a little weird but has other advantages in terms of accessibility and automation.
- Note: As part of this issue, we're also tackling some test improvements for offline simulation and predicate/priority forensics in scheduler_perf.

In summary:

I've coupled these so that test coverage can benefit from a forensics module - test coverage would be easier if there was a structured way to pull forensics data.  In fact, scheduler_perf could directly analyze the struct for this data during its run - which would solve the concrete issue identified above regarding the polling problem.

So to complete this issue, we need:

- [ ] To allow scheduler_perf to exersize the entire spectrum of the algorithm via pluggable predicates and priorities.

- [ ] To allow the mainline scheduler to export summary data i.e. via a REST endpoint, or text file..... which summarizes, without loss of information, the important events, including those which might otherwise lead to large-linear/exponential amount of logging.

**Comment by ncdc** [comment]
@jayunit100 would https://github.com/kubernetes-incubator/cluster-capacity be useful for this?

**Comment by jayunit100** [comment]
I think the above is an example of an app that could use the generic forensics utilities.

**Comment by jayunit100** [comment]
cc @kubernetes/sig-scheduling-feature-requests

**Comment by jayunit100** [comment]
cc @ConnorDoyle

**Comment by timothysc** [comment]
I'd rather see opentracing leveraged throughout the project for such items.  The problem with 1-off roll ups is it becomes unique for each component.  

I should have enough time this cycle to jig the scheduler for tracing.

**Comment by smarterclayton** [comment]
Yeah, would much rather we attack holistic tracing.  +1 from me.

On Wed, Apr 12, 2017 at 1:24 PM, Timothy St. Clair <notifications@github.com

**Comment by jayunit100** [comment]
@timothysc can we create a separate issue for the tracing stuff ? If so we can help test, verify its use for our needs also.  Then i will isolate this issue to just the instrumentation of the predicate/priority simulations.

**Comment by ncdc** [comment]
https://github.com/kubernetes/kubernetes/issues/26507 ?

**Comment by timothysc** [comment]
@ncdc yes.  

@jayunit100  https://youtu.be/NyySNe6Rr_g

**Comment by jayunit100** [comment]
I think this is perfect  "separate what you instrument from what you collect, turn things on and off" http://opentracing.io/documentation/pages/quick-start.html .  Thats exactly what i was hoping for.  I will rebrand the scheduler perf stuff in a separate issue.

**Comment by jayunit100** [comment]
Per https://github.com/kubernetes/kubernetes/issues/44342, i wonder wether we should revisit this at some point.
